### PR TITLE
Support more shells

### LINKED ISSUE #135: Support for more shells
1. besides `tcsh` and `bash`, #133 makes it clear we need to support `sh` as well
2. Can we figure out remote shell from `$SHELL` (and raise exception mandating use of `--shell` if that variable is undefined)
3. one possible path forward would be to create a parent class that contains wrappers for running commands and then shell-specific child classes with correct functionality (e.g. `sh -l -c "[run this command]"` for `sh` and `bash` but `sh -c "[run this command]"` for `tcsh`; also `'{command} > {self.log_file} 2>&1'` for `sh` and `bash` but `'{command} >& {self.log_file}'` for `tcsh`)
4. May also need template for expanding variables (e.g. `hostname` in `'jupyter lab --no-browser --ip=\$(hostname -f)'`)

Ideally this would let us set up simple tests launching commands other than `jupyter` so that we can use Github Actions to run these tests and make sure we are supporting as many shells as possible (and that future PRs don't break existing support)

### PR DESCRIPTION
**Pull request by andersy005**
- [x] Detect user's default shell
- [x] Fix major differences (redirect stdout and stderr to file, command substitution) among shells
- [x] More tests 
- [x] Closes #135 
- [x] Towards #70

### CODE REVIEW

**Comment by andersy005** [comment]
@mnlevy1981, I tried addressing some of the differences in shells. However, there are some corner cases (especially for tcsh/csh shells) I haven't tested so far. When you get a chance, please take a look and maybe try this out on a real machine and see how it goes.

**Comment by andersy005** [comment]
@mnlevy1981, feel free to push additional changes to this branch

File: tests/test_core.py [hunk 1]
```diff
@@ -93,27 +28,83 @@ def test_connection(runner):
...
+    out = runner._parse_log_file()
+    assert out == {
+        'hostname': 'eniac01',
+        'port': '59628',
+        'token': 'Loremipsumdolorsitamet',
+        'url': 'http://eniac01:59628/?token=Loremipsumdolorsitamet',
+    }
+
+
+@pytest.mark.skip(reason='Unable to diagnose CI issue causing this to fail')
```
Comment (andersy005) [inline tests/test_core.py:83]: This test keeps failing and I don't know why... We can address this in a separate PR

**Review by mnlevy1981 — COMMENTED** [review]
Looking at the output, the `test_conda_activate_cmd_error` is passing because it is producing an error, but I don't think this is the error we want:

```
tests/test_core.py::test_conda_activate_cmd_error[bash] ******* Running jupyter sanity checks (ensuring `jupyter` is in `$PATH`) *******
bash -c "source activate DOES_NOT_EXIST && command -v jupyter"
bash: activate: No such file or directory
 
`source activate` failed. Trying `conda activate`...
bash -c "conda activate DOES_NOT_EXIST && command -v jupyter"

CommandNotFoundError: Your shell has not been properly configured to use 'conda activate'.
To initialize your shell, run

    $ conda init <SHELL_NAME>

Currently supported shells are:
  - bash
  - fish
  - tcsh
  - xonsh
  - zsh
  - powershell

See 'conda init --help' for more information and options.

IMPORTANT: You may need to close and restart your shell after running 'conda init'.


 
PASSED
```

For `bash`, do we want to be using `sh -l -c {command}` instead of just `sh -c`? I'm pretty sure the existing code on `main` uses `-l`.

**Comment by andersy005** [comment]
@mnlevy1981, I update the PR accordingly... Thank you for your help! It turns out that there are some caveats with how shell integration works on GitHub actions: https://github.com/conda-incubator/setup-miniconda#important. Unfortunately, some tests keep failing on GitHub actions due to these caveats. So, I'm going to skip those tests for the time being.

**Comment by andersy005** [comment]
I'm satisfied with the [latest test coverage report](https://codecov.io/gh/ncar-xdev/jupyter-forward/tree/968887675a911c83733f3690702929d42b77a016/jupyter_forward) (Up from 52%)

Screen Shot 2022-02-24 at 7 57 36 PM

@mnlevy1981, let me know if you have any additional feedback.

**Review by mnlevy1981 — APPROVED** [review]
I pushed my last set of fixes with [f61fbbf](https://github.com/ncar-xdev/jupyter-forward/pull/138/commits/f61fbbf9bdde0bb9ab0a1ee1077516de035d66b1); this PR looks good to me!

**Review by kmpaul — APPROVED** [review]
This looks really cool, @andersy005  and @mnlevy1981!  Thanks for your work on this!
